Core/clim-basic/ports.lisp

```diff
@@ -171,107 +176,119 @@
-           (when (and sheet-t sheet-b)
-             (do ((s sheet-b (sheet-parent s)))
-                 ((or (null s) (graftp s) (eq s sheet-t)))
-               (let ((new-event (shallow-copy-object event)))
-                 (get-pointer-position (s event)
-                   (setf (slot-value new-event 'x) x
-                         (slot-value new-event 'y) y
-                         (slot-value new-event 'sheet) s))
-                 (change-class new-event 'pointer-exit-event)
-                 (dispatch-event s new-event))))))
-    (let* ((pointer-sheet (get-pointer-event-sheet (event-sheet event) event))
-           (old-pointer-sheet (or (port-pointer-sheet port) (event-sheet event)))
-           (common-sheet (sheet-common-ancestor old-pointer-sheet pointer-sheet)))
-      (distribute-exit-events old-pointer-sheet common-sheet event)
-      (distribute-enter-events pointer-sheet common-sheet event)
-      (setf (port-pointer-sheet port) pointer-sheet)
-      ;; set the pointer cursor
-      (when pointer-sheet
-        (let ((pointer-cursor (sheet-pointer-cursor pointer-sheet)))
-	  (unless (eql (port-lookup-current-pointer-cursor port (event-sheet event))
-		       pointer-cursor)
-	    (set-sheet-pointer-cursor port (event-sheet event) pointer-cursor))))))
-  ;; Do not send the event twice.
-  (unless (typep event '(or pointer-enter-event pointer-exit-event))
-    (call-next-method)))
-
-;;; This method is necessary to distribute events for sheets which do
-;;; not have a mirror. In this case we find the innermost sheet at
-;;; event's position and dispatch event's shallow copy to this
-;;; sheet. -- jd 2019-08-05
+                    (t (setf sheet-a (sheet-parent sheet-a)))))))
+    (let ((common-ancestor (sheet-common-ancestor old-pointer-sheet
+                                                  new-pointer-sheet)))
+      ;; distribute exit events (innermost first)
+      (when (and common-ancestor old-pointer-sheet)
+        (do ((s old-pointer-sheet (sheet-parent s)))
+            ((or (null s) (graftp s) (eq s common-ancestor)))
+          (dispatch-event-copy s event 'pointer-exit-event)))
+      ;; distribute enter events (innermost last)
+      (dolist (s (do ((s new-pointer-sheet (sheet-parent s))
+                      (lis nil))
+                     ((or (null s) (graftp s) (eq s common-ancestor)) lis)
+                   (push s lis)))
+        (dispatch-event-copy s event 'pointer-enter-event)))))
+
+(defmethod distribute-event ((port basic-port) event)
+  (dispatch-event (event-sheet event) event))
+
+;;; In the most general case we can't tell whenever all sheets are
+;;; mirrored or not. So a default method for pointer-events operates
+;;; under the assumption that we must deliver events to sheets which
+;;; doesn't have a mirror and that the sheet grabbing and pressing is
+;;; not implemented by the port (we emulate it). -- jd 2019-08-21
 (defmethod distribute-event ((port basic-port) (event pointer-event))
```

Comment:
The commit message says

> [...] will behave slightly different, namely when pointer will reach different mirror our "pressed sheet" will lose focus and it won't receive events.

If I understand the new intended behavior correctly, I'm not sure it is good. Consider a frame with a scroll-bar on its right border and a larger frame "below" it (region of the first sheet is inside the region of the second sheet in graft coordinates). If I press and hold the pointer button to move the scroll-bar, I can let the pointer wander off the scroll-bar to the left (further inside the top frame), but if the pointer moves to the right onto the bottom frame (button still pressed), the scroll-bar will stop responding to pointer motion.

Reply:
it is the other way around. if you use default clx mechanisms you will lose focus, in this changed behavior which I've introduced focus will stay. Should I phrase the commit message differently?

Reply:
fixed in a way that we do not distribute event if it is a release button for one of the scrollings (so I've preserved the intent after all)